Improved control over crash report "title" (#51)

ssawchenko opened the pull request:
### Related Issue: #51 


### Summary of Problem:
When Sentry support was added, we bumped into a situation where the report being made had a title that included the extra data string; this caused the list of reports to become very hard to read.
![image](https://user-images.githubusercontent.com/5217641/114778600-ea984600-9d29-11eb-99f7-083ec07c5210.png)

This PR fixes those formatting issues and forces extra data to be written as logs/breadcrumbs on crash reports

### Proposed Solution:
* Created a class that can encapsulate the message, original throwable and any extra data that should be written to the logs.
* Each Destination is responsible for determining how best to include all of these objects.
* Crash reporting Destinations no longer include the "extra" (Redactable) data in the log titles.

### Testing Steps:
I checked the output of each Destination to be sure it looked correct:

**Sentry:**
![Screen Shot 2021-04-14 at 1 30 54 PM](https://user-images.githubusercontent.com/5217641/114778958-58dd0880-9d2a-11eb-9ec8-394a22ccd9a1.png)
Breadcrumbs showing messages and extra data:
![image](https://user-images.githubusercontent.com/5217641/114920308-27257980-9dde-11eb-972e-58a397b249c7.png)

**Crashlytics (although this is now deprecated in favour of Sentry):**
![Screen Shot 2021-04-14 at 1 28 07 PM](https://user-images.githubusercontent.com/5217641/114778990-61cdda00-9d2a-11eb-882a-aadc561c51cf.png)
Logs showing messages and extra data:
![image](https://user-images.githubusercontent.com/5217641/114779141-96419600-9d2a-11eb-8a17-cd81d46b5e06.png)

**File:** 
![image](https://user-images.githubusercontent.com/5217641/114921727-bb441080-9ddf-11eb-91ff-2eb0b4155b32.png)

**Console:**
![image](https://user-images.githubusercontent.com/5217641/114921887-e464a100-9ddf-11eb-9404-2d7c73a9f094.png)

jacobminer commented on app/src/main/java/com/steamclock/steamclogsample/MainActivity.kt:
```diff
@@ -48,45 +55,98 @@ class MainActivity : AppCompatActivity() {
+
+        // Should create tickets
+        //
+        // NOTE, Sentry seems to apply some grouping logic such that the *same* Throwable thrown
```
We might need a way to override this behaviour later? I know iOS has the ability to send in a non-fatal as a "user report", which always needs to generate its own issue.

ssawchenko replied:
Ok, I'll create a ticket for that. I think that given the changes I made here it should *hopefully* be easy to support as long as the "Message" used to create the user report is unique. We would use `Sentry.captureMessage()` instead of `Sentry. captureException()` which I think would avoid this "grouping" case.

ssawchenko replied:
https://github.com/steamclock/steamclog-android/issues/56

jacobminer commented on steamclog/src/main/java/com/steamclock/steamclog/Destinations.kt:
```diff
@@ -36,7 +38,7 @@ internal class CrashlyticsDestination : Timber.Tree() {
             fun with(message: String): NonFatalException {
                 val stackTrace = Thread.currentThread().stackTrace
-                val numToRemove = 10 // Move down the stack past Timber and Steamclog calls.
+                val numToRemove = 12 // Move down the stack past Timber and Steamclog calls.
```
Can we file a ticket for adding unit tests for verifying this is working as expected? If we forget to update this number and push it to a client app, it could mess up our logging pretty badly.

ssawchenko replied:
https://github.com/steamclock/steamclog-android/issues/57

ssawchenko replied:
NonFatalException was a work around to support Crashyltics quirks, so once we have fully removed support for it we can remove this as well.

jacobminer commented on steamclog/src/main/java/com/steamclock/steamclog/Destinations.kt:
```diff
@@ -48,12 +50,29 @@ internal class CrashlyticsDestination : Timber.Tree() {
     }
 
     override fun log(priority: Int, tag: String?, message: String, throwable: Throwable?) {
+        val wrapper = throwable as? SteamclogThrowableWrapper
```
Are there any cases where we're expecting `throwable` to not be a `SteamclogThrowableWrapper`?

ssawchenko replied:
With this PR we should only be getting `SteamclogThrowableWrapper`

jacobminer replied:
Should we use `as` instead of `as?` in this case then?

ssawchenko replied:
Yeah, I was just thinking about this a little more. Maybe this should be slightly more flexible - I would hate for a crash to happen somewhere because in the app someone uses `Timber` instead of `sclog` by accident.

ssawchenko replied:
`SteamclogThrowableWrapper.from` now enforces that we will always have a valid object moving forward.

ssawchenko commented:
Ok @jacobminer after the above comments and some more thought I expanded this PR a little to fix some more things:

1. Each destination will now call `SteamclogThrowableWrapper.from(Throwable)` which will convert the Throwable to our wrapper if required. This should address the nullability issue in your comments.
2. I added `generateSimpleLogMessage` and `generateFullLogMessage` methods to make the message formatting consistent; wasn't one of your comments, but I noticed while changing things that each Destination was applying their own formatting which would make it much harder to unit test later.
3. I added some testing in the `createCustomStackElementTag` method on debug builds; since the stack trace is highly dependant on the call location this method is hard to unit test right now - I figured adding in tests on debug was a decent stop gap. If you run the sample app with an incorrect index in this method the following crash is now seen:
![Screen Shot 2021-04-15 at 11 13 14 AM](https://user-images.githubusercontent.com/5217641/114919886-a9fa0480-9ddd-11eb-9604-cc2fbe7acff7.png)
4. In order to support these tests I had to make a differentiation between logs made from outside the Steamclog library and those made within; to do this I created the `logInternal` method.

jacobminer commented on steamclog/src/main/java/com/steamclock/steamclog/Destinations.kt:
```diff
@@ -145,49 +163,45 @@ internal class ConsoleDestination: Timber.DebugTree() {
-            val logStr = "$logTimeStamp $appId[$processId:$threadName] [$priority] [$tag] > $message \r\n"
-            // If file created or exists save logs
-            getExternalFile()?.let { file -> file.appendText(logStr) }
+            getExternalFile()?.let { file -> file.appendText("$message\r\n") }
```
Looks like this wasn't added in this PR, but do you know if this will automatically close the file after writing is finished?

ssawchenko replied:
Hrm. Doesn't look like I'm closing it, but it is still working. I've created https://github.com/steamclock/steamclog-android/issues/58 to follow up on it and make sure it's handled correctly.
